CSysMatrix cleanup and performance improvements

pcarruscag opened the pull request:
## Proposed Changes
I made this in micro steps and the commit messages are fairly detailed, so here I will give only the highlights. Every change is mathematically equivalent.
I claim performance improvements based on the type of cases I run (steady RANS implicit, elasticity), results may vary, so please give this a try.

**Cleanup**
- Iterative smoothers (Jacobi, ILU, etc.) are now implemented in CSysSolve as one generic method with the same interface as the Krylov solvers.
- Dead code was removed from CSysMatrix (as requested in #653).
- Duplicated code, mostly dealing with block-block/vector multiplication, was merged.
- Row/column elimination tasks implemented in CSysMatrix (as requested in #658).
- Linelet uses fewer working variables.
- Make a lot of methods private (the diff of matrix_structure.hpp is not going to be pretty...).

**Performance**
- Inlined small methods.
- No calls to "Get/SetBlock" when the block location is already known (helps ILU and LU_SGS).
- Block inversion done in one go instead of multiple Gaussian eliminations (which used to require repeated upper matrix transformations) (helps Jacobi, Linelet, and ILU).
- Inverted ILU diagonal blocks are stored during "build" for use in "compute".

**Bugs**
- Linelet preconditioner was not working for multiple wall markers.

**ToDo**
- I will try to make the MKL optimizations work for the discrete adjoint (I say try because it may require too much creativity with the templates...).
- Get some benchmarks (I don't want you to just take my word for it).

## Related Work
#650, #653, #658

## PR Checklist
- [X] I am submitting my contribution to the develop branch.
- [X] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags).
- [X] My contribution is commented and consistent with SU2 style.
- [X] I have added a test case that demonstrates my contribution, if necessary.

economon commented on SU2_CFD/src/solver_direct_elasticity.cpp:
```diff
@@ -2009,82 +2009,35 @@ void CFEASolver::BC_Clamped(CGeometry *geometry, CSolver **solver_container, CNu
+
+    LinSysSol.SetBlock(iPoint, Solution);
+    LinSysReact.SetBlock(iPoint, Solution);
+    Jacobian.EnforceSolutionAtNode(iPoint, Solution, LinSysRes);
```
Nice - thanks for following up on this w/ the EnforeSolutionAtNode() method

economon commented on Common/src/matrix_structure.cpp:
```diff
@@ -1457,327 +1049,35 @@ void CSysMatrix<ScalarType>::MatrixVectorProductTransposed(const CSysVector<Scal
-}
-
-template<class ScalarType>
-unsigned long CSysMatrix<ScalarType>::Jacobi_Smoother(const CSysVector<ScalarType> & b, CSysVector<ScalarType> & x, CMatrixVectorProduct<ScalarType> & mat_vec, ScalarType tol, unsigned long m, ScalarType *residual, bool monitoring, CGeometry *geometry, CConfig *config) {
```
Thanks for collapsing all of my old smoothers into a unified implementation. Can some of the comments here be preserved in the new implementations under the Compute*Preconditioner() routines? I recall adding many of them to explain step by step what each of the smoothers were doing...

pcarruscag replied:
Sure can, I'll see if there were comments in the smoothers that are not in the preconditioner.

economon commented on Common/src/linear_solvers_structure.cpp:
```diff
@@ -651,6 +652,95 @@ unsigned long CSysSolve<ScalarType>::BCGSTAB_LinSolver(const CSysVector<ScalarTy
+                                                        ScalarType tol, unsigned long m, ScalarType *residual, bool monitoring, CConfig *config) {
+
+  int rank = SU2_MPI::GetRank();
+  ScalarType norm_r = 0.0, norm0 = 0.0, omega = 1.0; // hardcoded relaxation factor
```
Have you investigated the effects of the relaxation factor? When I first put this in, I included a variable for it, but I am not sure I tested it much. Would be interested to see.

Also might be worth exposing this to the config and/or making it consistent with the relaxation factors appearing in other linear solvers.

pcarruscag replied:
Not in SU2, but the theory should be that if the preconditioner does not approximate the inverse of the matrix well you need smaller relaxation, you would want the relaxation that minimizes the spectral radius but that is not easy to compute. I also do not know what is the effect on the "quality" of the smoothing...

I can expose it to the config but I'm not too keen on running parameters tests on this.
By consistent with the other solvers do you mean computing it every iteration with some formula (e.g. such that the norm of the residual is minimized)? With this I have played with and it very easily stalls (omega=0).

Thank you for starting the review this quickly by the way.

economon replied:
Aok, no need to test rigorously. I would be interested to have this exposed to the config though.

As for consistency with other relaxation factors, it is probably beyond this PR, and it can wait for later. Basically, what I meant is that it would be nice to see a more thought-out strategy for relaxation factors that affect the linear solvers. The current approach seems disjointed, as we have RELAXATION_FACTOR_FLOW that multiplies the output increment of the linear solve in the FV flow solver, for instance, and then other potential relaxation factors inside the linear solver, such as the one in question, and maybe other physics solvers that don't apply relaxation at all in the linear solver. Just food for thought for now.

economon commented on Common/include/matrix_structure.hpp:
```diff
@@ -280,127 +248,90 @@ class CSysMatrix {
-   * \param[in] matrix
-   * \param[in] vector
-   * \param[out] product
+   * \brief Performs the product of i-th row of the lower part of a sparse matrix by a vector.
```
Thanks for attention to detail in the comments

pcarruscag replied:
Borderline ocd... but oh well.

economon commented on Common/src/matrix_structure.cpp:
```diff
@@ -1457,327 +1049,35 @@ void CSysMatrix<ScalarType>::MatrixVectorProductTransposed(const CSysVector<Scal
-  }
+void CSysMatrix<ScalarType>::BuildILUPreconditioner(bool transposed) {
   
-  //  ScalarType Det, **Matrix, **CoFactor;
```
Thanks for the clean up here as requested too

pcarruscag commented on Common/src/linear_solvers_structure.cpp:
```diff
@@ -738,7 +828,17 @@ unsigned long CSysSolve<ScalarType>::Solve(CSysMatrix<ScalarType> & Jacobian, CS
+  /*--- For smoother-type solvers set the associated preconditioner type. ---*/
+
+  switch (KindSolver) {
+    case SMOOTHER_LINELET: KindPrecond = LINELET; break;
```
Should we just have KindSolver SMOOTHER in the config now?

economon replied:
I like this idea, yes!

pcarruscag replied:
Done

pcarruscag commented on Common/include/matrix_structure.inl:
```diff
@@ -160,25 +175,364 @@ template<class ScalarType>
+}
+
+#define __MATVECPROD_SIGNATURE__(TYPE,NAME) \
+inline void CSysMatrix<TYPE>::NAME(const TYPE *matrix, const TYPE *vector, TYPE *product)
```
I hope this does not look too bad, but this MKL stuff does work very well for the discrete adjoint (maybe 10% faster than native code).

talbring commented:
Thanks @pcarruscag! This is a nice cleanup! :+1: 

So while you are already at it, you can already separate it in folders/files and remove the *inl (if all agree with these changes). Maybe adding a linear algebra subfolder makes sense.

pcarruscag commented:
[Here](https://github.com/su2code/SU2/files/3302370/benchmark.zip) are some comparisons for a case where the linear solver does little work (so not much benefit from this PR) and another where the linear solver has to work harder (18 iterations per iteration on average) and this makes a substantial difference.
I have started updated the residuals of things that seem consequence of round-off error, only one case requires closer inspection where things change several orders of magnitude (for the best).

pcarruscag commented:
Thanks @talbring,
I moved the preconditioner and matrix-vector product wrapper classes to separate files, these are so light weight that I was thinking of leaving them in one file.
I also moved some inlines to the hpp but I kept the private inlines in the inl file, these are only needed in the cpp and so by including the inl from the cpp (instead of bottom of hpp) we might avoid triggering recompilation of more units when working on implementation details of CSysMatrix.
Finally I would like to move/rename the larger files on a separate PR, that way it will be easier to track changes.
What do you think?

pcarruscag commented:
The case with large change of residuals (3 orders) is the rotating cylinders case (sliding interface).
With this PR I get the following flow field at the last iteration:
![image](https://user-images.githubusercontent.com/38071223/59783874-0ca11880-92b9-11e9-840e-20fd87ec17e9.png)
With develop I get this one at time iter 2:
![image](https://user-images.githubusercontent.com/38071223/59783928-2cd0d780-92b9-11e9-90bd-9155dd2a6b4e.png)
And I think it is fair to say the case was actually blowing up before:
![image](https://user-images.githubusercontent.com/38071223/59784008-5984ef00-92b9-11e9-89e0-d78eb0d76b90.png)
If I change the linear preconditioner to the ILU the develop results are basically the same as with this PR. Why does the previous LU_SGS diverge and the current doesn't? No idea, maybe for a different limit condition it would go the other way...

talbring commented:
@pcarruscag Good that you mention it. I saw exactly the same behavior while working on #715. If I disable grid movement for the first zone (where grid velocity is 0 anyway because there is no rotation) it works. In my opinion having grid movement disabled or enabled and specifying zero movement should result in exactly the same convergence behavior ...

pcarruscag commented:
I tested the MKL integration with the discrete adjoint, and everything looks ok, turns out to be only about 5% faster on a per iteration basis (i.e. excluding recording times). I had to grab some work from my other ongoing PR's to run a case with reasonable CFL settings so I am not going to upload files for this test.

@economon , @talbring , if you do not mind me moving the files with significant changes on a separate PR, I think this is ready to go.

economon commented:
@talbring : as you know, the difference between disabled and active grid movement with 0 velocity is that the former case does not even allocate the memory for the grid velocity at each node, and many conditionals checking for grid movement throughout the solver (fluxes, BCs) are avoided. This was to save memory and effort when grid motion is not needed, however, maybe we need to now change the strategy for multizone problems which may have both fixed and moving zones (perhaps always active with 0 as default). 

I am a little surprised they are not the same as well, but somewhere in the code path there must be an issue with this.. my guess is something related to BC_Fluid_Interface() or the transfer structure when grid movement is active on both sides but has a value of 0 on one of the interfaces.

talbring commented:
@pcarruscag For me its fine if you do it in a separate PR.

@economon I also assume that is has something to do with how the interfaces are handled. We definitely have to check it. For now in PR #715 I enabled grid movement (in the config) also in fixed zones.

pcarruscag commented:
If that case is so sensitive to be affected by the changes in this PR (it started converging without any change of options), it does not surprise me that other sources of numerical noise can have a similar influence.

economon commented:
Ok for me to wait for another PR to break the other files.

Would be good to get some additional reviews on this one from the community if folks have some time.

pcarruscag commented on Common/src/matrix_structure.cpp:
```diff
@@ -1656,9 +886,9 @@ void CSysMatrix<ScalarType>::BuildILUPreconditioner(bool transposed) {
           
-          if ((kPoint >= jPoint) && (jPoint < (long)nPointDomain)) {
+          if (kPoint > jPoint) {
             
```
This was not a bug, but we overwrite the ij block completely so it is not necessary to update it here.

pcarruscag commented:
@economon no one seems to have time, let's get going please.

EduardoMolina commented on .travis.yml:
```diff
@@ -81,7 +81,7 @@ install:
 before_script:
     # Get the test cases
-    - git clone -b develop https://github.com/su2code/TestCases.git ./TestData
+    - git clone -b feature_refactor_lin_solvers https://github.com/su2code/TestCases.git ./TestData
```
Please revert before we merge.

EduardoMolina commented on Common/include/linear_solvers_structure.hpp:
```diff
@@ -242,19 +248,35 @@ class CSysSolve {
+  unsigned long Smoother_LinSolver(const VectorType & b, VectorType & x, ProductType & mat_vec,
+                                   PrecondType & precond, ScalarType tol, unsigned long m,
+                                   ScalarType *residual, bool monitoring, CConfig *config);
+  
```
Thanks for the comments.

pcarruscag commented:
Thank you @WallyMaier and @EduardoMolina for the reviews.
